feat: Add metadata support to custom detectors

effortlessdevsec opened the pull request:
# Feature Request: Add Metadata Support to Custom Detectors

## Summary
Add support for a `metadata` field in custom detector configurations that automatically populates the `ExtraData` field in detector results. This would allow users to attach custom key-value metadata directly in the YAML configuration without requiring code changes.

## Use Case
Currently, custom detectors can only set metadata programmatically in the code. Users should be able to define metadata directly in their custom detector YAML configuration files, which would then be automatically included in the `ExtraData` field of all results from that detector.

This would be useful for:
- Adding environment tags (e.g., `environment: production`)
- Adding team ownership information (e.g., `team: security`)
- Adding severity levels (e.g., `severity: high`)
- Adding rotation guides or documentation links
- Adding any custom contextual information that should be associated with detected secrets

## Proposed Implementation

### 1. Proto Definition Update
Add a `metadata` field to the `CustomRegex` message in `proto/custom_detectors.proto`:

```protobuf
message CustomRegex {
  // ... existing fields ...
  map<string, string> metadata = 12;
}
```

### 2. Code Implementation
Update `pkg/custom_detectors/custom_detectors.go` to copy metadata from the detector configuration to `ExtraData` when creating results:

```go
// In createResults function
if metadata := c.GetMetadata(); metadata != nil {
    for key, value := range metadata {
        result.ExtraData[key] = value
    }
}
```

## Example Usage

### YAML Configuration
```yaml
detectors:
- name: my-api-key-detector
  keywords:
  - api
  - key
  regex:
    api_key: "your-regex-here"
  metadata:
    environment: "production"
    team: "security"
    severity: "high"
    rotation_guide: "https://example.com/rotate-api-keys"
    custom_field: "any value"
```

### Result
All results from this detector would automatically include the metadata in `ExtraData`:
```json
{
  "DetectorName": "my-api-key-detector",
  "ExtraData": {
    "name": "my-api-key-detector",
    "environment": "production",
    "team": "security",
    "severity": "high",
    "rotation_guide": "https://example.com/rotate-api-keys",
    "custom_field": "any value"
  }
}
```

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new user-configurable field to the custom detector protobuf and threads it into emitted results; low algorithmic risk but changes output shape and could impact downstream consumers relying on `ExtraData` contents.
> 
> **Overview**
> Custom detector configs now accept a `metadata` map (proto field `CustomRegex.metadata`) and generated pb/validate code has been updated accordingly.
> 
> `CustomRegexWebhook` results now copy configured metadata into `Result.ExtraData` when creating results, while ensuring `ExtraData` is non-nil when adding the detector `name` and when storing webhook verification `response`. New tests cover empty vs populated metadata being present in emitted `ExtraData`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8dd1ccceec78de91526c7e1a7751d336bf50e03e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

effortlessdevsec committed:
Update custom_detectors.proto

effortlessdevsec committed:
Update custom_detectors.go

kashifkhan0771 commented:
You need to run the command `make protos` after making changes in the `.proto` file.

camgunz reviewed (changes requested):
This is a good idea! Requesting changes for:
* running `make protos` as Kashif suggested
* adding at least 2 tests:
    * an empty `metadata` field adds nothing to `ExtraData`
    * a populated `metadata` field adds expected values to `ExtraData`

You could also move the allocation of the map inside the `if metadata...` block (maybe also add a `len(metadata) > 0` check?), but I don't think it makes a huge difference.

effortlessdevsec committed:
Merge branch 'trufflesecurity:main' into main

Commit:
Add metadata support to custom detectors

- Add metadata field to CustomRegex proto message
- Copy metadata from detector config to ExtraData in results
- Only allocate ExtraData map when metadata exists (optimization)
- Add tests for empty and populated metadata scenarios
- Run make protos to regenerate proto code

effortlessdevsec committed:
Merge branch 'main' into main

Commit:
Fix test string to avoid gitleaks detection

Commit:
Merge remote main and resolve conflict - keep optimized metadata allocation
